lib/lists.nix

```diff
@@ -641,8 +677,19 @@ rec {
      Example:
        unique [ 3 2 3 4 ]
        => [ 3 2 4 ]
-   */
- unique = foldl' (acc: e: if elem e acc then acc else acc ++ [ e ]) [];
+  */
+  unique = foldl' (acc: e: if elem e acc then acc else acc ++ [ e ]) [];
+
+  /* Sort and remove duplicate elements from the list.
+     O(n) complexity on top of `sort`.
+
+     Type: fastUnique :: (a -> a -> bool) -> [a] -> [a]
+
+     Example:
+       fastUnique (a: b: a < b) [ 3 2 3 4 ]
+       => [ 2 3 4 ]
+  */
+  fastUnique = comparator: list: uniq (sort comparator list);
```

Comment:
I think we can have a more generic form by implementing this with
```nix
comparator: list: uniqBy (a: b: ! comparator a b) (sort comparator list)
```

This works because `uniqBy`'s predicate is only ever called on increasing list values, which are sorted already (aka `a <= b`). With the additional check `! a < b` we can then ensure that they're equal.

Reply:
I'm reminded of Rust's [`core::cmp::PartialEq`](https://doc.rust-lang.org/stable/core/cmp/trait.PartialEq.html) & [`core::cmp::PartialOrd`](https://doc.rust-lang.org/stable/core/cmp/trait.PartialOrd.html) traits. Using inverted `comparator` as a `uniqBy` argument implies that, letting `a < b = comparator a b`, if `a < b && !(b < a)` then `a == b`. This is not the case if `a` and `b` are partially ordered by `comparator` and not totally ordered. Nix doesn't have NaN to make its floats not totally ordered, but the user-supplied data ordered by user-supplied `comparator` might not be totally ordered. However, this total ordering requirement can be disclosed in documentation.

I'm more worried about a user-supplied `comparator` being substantially slower than a user-supplied `pred` would be. If user-supplied data doesn't follow Nix `==` equality, then forcing explicit fallback to `uniqBy pred (sort comparator list)` (which isn't much longer) makes it clearer how `uniqBy (a: b: !(comparator a b)) (sort comparator list)` would repeatedly run a potentially-expensive `comparator`.

Expensive `comparator`s might not be an issue in practice, though, and I defer here to those familiar with more of Nixpkgs.

Alternatively, `fastUnique` could be defined as `fastUnique = list: uniq (sort lessThan list);`, and we could force manual expansion for any other usage. I don't see much of a benefit to this though, given that `builtins.sort` takes a `comparator`. I'd like to mirror `builtins.sort` here. For comparison, `builtins.elem` does not take a `pred`, and `uniq` & `unique` reflect this. If the pattern of `uniq` → `uniqBy` was followed, there would be `fastUniqueBy = pred: comparator: list: uniqBy pred (sort comparator list);`. `fastUniqueBy` could shorten that earlier explicit expansion to `fastUniqueBy (a: b: !(comparator a b)) comparator list`, but I don't think defining this adds much?